### PR TITLE
Mongoid criteria refactor

### PR DESCRIPTION
**Pull request by asgerb**
A proposal for a refactor of the mongoid criteria.

I’ve moved it from `Mongoid::Criterion::Scrollable` to `Mongoid::Criteria::Scrollable` since `Criterion` was removed in version 4 of mongoid.

I’ve changed the `criteria = self` to `criteria = dup`. This allowed me to use bang methods, which provided a benefit when setting the default sort option. 

Otherwise I haven’t changed anything substantial about the code, but have tried to make it more readable. This means I’ve moved the named variables to method calls. I guess one downside of this is performance?

Perhaps some of the decisions are going a bit too far, looking forward to your thoughts!

I’ve also added a spec which checks that the base criteria does not change, but have kept the `cursor_criteria = criteria.dup` from the original code. While I don’t think the `dup` was needed before, it is needed now, since the named variables have been removed.

### CODE REVIEW

File: lib/mongoid/criteria/scrollable.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            yield record, cursor_from_record(record, cursor_options)
+          end
+        else
+          cursor_criteria
+        end
+      end
+
+      private
+
+      def raise_multiple_sort_fields_error
```
Comment (dblock) [inline lib/mongoid/criteria/scrollable.rb]: Should be a bang method I think, so `raise_multiple_sort_fields_error!`, for clarity that it fails loudly.

File: lib/mongoid/criteria/scrollable.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        raise Mongoid::Scroll::Errors::MultipleSortFieldsError.new(sort: criteria.options.sort)
+      end
+
+      def multiple_sort_fields?
+        options.sort && options.sort.keys.size != 1
+      end
+
+      def no_sort_option?
```
Comment (dblock) [inline lib/mongoid/criteria/scrollable.rb]: Maybe flip it to `sort_option?`?

File: lib/mongoid/criteria/scrollable.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        Mongoid::Scroll::Cursor.from_record(record, cursor_options)
+      end
+
+      def scroll_field_type(criteria)
+        scroll_field = scroll_field(criteria)
+        field = criteria.klass.fields[scroll_field.to_s]
+        field.foreign_key? && field.object_id_field? ? bson_type : field.type
+      end
+
+      def bson_type
```
Comment (dblock) [inline lib/mongoid/criteria/scrollable.rb]: You should add this to mongoid-compatibility, maybe `Mongoid::Compatibility::ObjectId#bson_type`. Obviously fine for now here.

**Comment by dblock** [comment]
@asgerb Are you interested in joining the maintainers group? Maybe make the next release? (what's your rubygems email/name if so).

**Comment by asgerb** [comment]
Hey @dblock, thanks for the quick merge, happy that this was useful 🙂
I'll take a look at your proposals and make another PR.
About joining as a maintainer, I'm afraid I have zero experience as a maintainer, so I'm not sure I'm the best fit? 😉 I don't even have a rubygems login, and never released a gem heh.

**Comment by dblock** [comment]
I think you'll do fine @asgerb so this could be your first! You seem to know what you're doing otherwise :) Get a Rubygems login and let me know. I can help you avoid doing anything stupid and we have instructions for things like releasing. Otherwise it's just reviewing other people's PRs and helping in any way you can.

**Comment by asgerb** [comment]
Sounds good, I'm up for it 🙂 I'll sign up on rubygems then!

**Comment by asgerb** [comment]
OK I'm on Rubygems with a@asgerbehnckejacobsen.dk 🙂

**Comment by dblock** [comment]
All added @asgerb. Welcome!

**Pull request merged**